### PR TITLE
ENH: Add GetNumberOfPixels to Images

### PR DESCRIPTION
**Pull request by hjmjohnson**
This can assist with integration with other libraries,
and in simple cases it allows for using the std algorithms
more easily.

float * imgdata = img.GetDataAsFloat();
float total = std::accumulate(imgdata, imgdata+img.GetNumberOfPixels, 0);

### CODE REVIEW

File: Code/Common/src/sitkPimpleImageBase.hxx [hunk 1]
```diff
@@ -261,18 +261,30 @@ namespace itk
...
       {
-        typename ImageType::RegionType largestRegion = this->m_Image->GetLargestPossibleRegion();
+        const typename ImageType::RegionType & largestRegion = this->m_Image->GetLargestPossibleRegion();
         std::vector<unsigned int> size( ImageType::ImageDimension );
 
         return sitkITKVectorToSTL<unsigned int>( largestRegion.GetSize() );
       }
 
+    virtual unsigned int GetNumberOfPixels( void ) const
+      {
```
Comment (blowekamp) [inline Code/Common/src/sitkPimpleImageBase.hxx]: Please use ImageRegion::GetNumberOfPixels

File: Code/Common/include/sitkImage.h [hunk 2]
```diff
@@ -188,6 +188,7 @@ namespace simple
     std::vector< double > TransformContinuousIndexToPhysicalPoint( const std::vector< double > &index) const;
 
     std::vector< unsigned int > GetSize( void ) const;
+    unsigned int GetNumberOfPixels( void ) const;
```
Comment (blowekamp) [inline Code/Common/include/sitkImage.h]: Please add documentation to clarify how this interacts with the number of components and for vector images.

**Comment by blowekamp** [comment]
This looks like a good idea.

The return type be int64_t or uint64_t type. Because not all languages ( Java ) support unsigned integers. (I'll get back to you on this).

The addition of "const ImageRegion" looks good. However, since I consider the ImageRegion type to be "small" since it only contains fixed arrays, I am not sure where there is any benefit to the "return to reference" added. Perhaps it still just looks little odd and I need to get use to it still.

File: Code/Common/include/sitkImage.h (hunk 2 above)
Comment (hjmjohnson) [inline Code/Common/include/sitkImage.h]: Done.

File: Code/Common/src/sitkPimpleImageBase.hxx (hunk 1 above)
Comment (hjmjohnson) [inline Code/Common/src/sitkPimpleImageBase.hxx]: Done.

**Comment by hjmjohnson** [comment]
@blowekamp I am changing the signature to uint64_t, but the other member functions are still using unsigned integers.  GetWidth, GetDepth, GetHeight, GetNumberOfComponents.  Will those also need to be changed?

**Comment by blowekamp** [comment]
No.

Indexes are 32-bits or unsigned int, but the total number of pixels should be uint64_t. IMHO, the indices should be the natural word of the system doing the computation where as the "size" or count should be the largest. But there is challenge of trying to provide a consistent interface across systems.

Currently, there is some inconsistent use of "unsigned int"  and "uint32_t" for the index types.

File: Code/Common/include/sitkImage.h [hunk 3]
```diff
@@ -149,6 +149,17 @@ namespace simple
      * number of components for each pixel is returned.
      */
     unsigned int GetNumberOfComponentsPerPixel( void ) const;
+    /** \brief Get the number of pixels in each image
```
Comment (blowekamp) [inline Code/Common/include/sitkImage.h]: Please add a new line before the Doxygen comment.

File: Code/Common/include/sitkImage.h [hunk 4]
```diff
@@ -149,6 +149,17 @@ namespace simple
...
+     *
+     *  Some pixels may have more than one component, so
+     *  GetNumberOfPixels should be multiplied by
+     *  GetNumberOfComponentsPerPixel in order to calculate the
+     *  the total number of values stored congituously for the image.
+     *
+     * The GetNumberOfPixels produces the same result
+     * as GetWidth()*GetHeight()*GetDepth().  This value
```
Comment (blowekamp) [inline Code/Common/include/sitkImage.h]: The above is incorrect. For 2D images as GetDepth() is 0, and SimpleITK also has the option to support 4D images too. And the last sentence appears to be unfinished.

**Comment by blowekamp** [comment]
Also a couple simple tests should be added to "Testing/Unit/sitkImageTest.cxx" under the Image.Transform block.

Please let me know if you need to to finish up this patch.

Thanks for the contributions!

**Comment by blowekamp** [comment]
This patch was updated and merged here: 2fcbf38189364d74b05acdbb05aedeeb1efcea1b

**Pull request closed without merging**